subsys: logging: Added File System backend log

m-syc opened the pull request:
This PR introduces a new backend based on littleFS. New backend can be turned on
in menuconfig by enabling `logging--> Enable FLASH backend`.

In init, there is created a new file with continuous numeration. When max size of file is reached,
system creates another one. There is a limit of log files used by backend too. There is a possibility to overwrite
old files or drop new ones. User can also change the name of log files and decide to mount a file system by own
or leave this obligation for the file system. All of them are available to set via Kconfig.

~~In this PR there are additional 2 tests: with auto mounted file system y backend and manual by user.
Both check if log file contains data logged via API.~~

There is a change in mcumgr sample too. This project could be used to test new backend.
The easiest way to download the log file is using this command:
`mcumgr --conntype serial --connstring='dev=/dev/ttyACM0,baud=115200' fs download /lfs/log.0000 ~/log.txt`.
I also updated documentation of mcumgr sample by adding tips how to download log file using `mcumgr` app.

Additionally there are added dependencies in CMakeLists file.

- [x] reworked for testing backends behavior, without mediation of `LOG`
- [x] fix monoethnic log files replacement
- [x] auto recovery after unexpected lost of the current log file in runtime.

nordic-krch commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +279,40 @@ config LOG_FRONTEND
+
+if LOG_BACKEND_FLASH
+
+config OVERWRITE_OLD_LOG_FILES
```
it's good to prefix kconfig options with module. It would be descriptive enough to call it `LOG_BACKEND_FLASH_OVERWRITE`

MSyc-NS replied:
Macro renamed to `LOG_BACKEND_FS_OVERWRITE` following suggestion from another comment.

nordic-krch commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +279,40 @@ config LOG_FRONTEND
 	  When enabled, logs are redirected to a custom frontend instead
 	  of being processed by the logger.
 
+config LOG_BACKEND_FLASH
```
I wonder if this backend should not be called `LOG_BACKEND_FS`  (or `LOG_BACKEND_FILESYSTEM`) since it is using file system API rather than flash API.

MSyc-NS replied:
Replaced by `LOG_BACKEND_FS` prefix.

nordic-krch commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +279,40 @@ config LOG_FRONTEND
+	  When enabled backend overwrites oldest log files.
+	  In other case, when memory is full, new messages are dropped.
+
+config LOG_FLASH_FILE_SIZE
```
i would prefix it with module name `LOG_BACKEND_FLASH_FILE_SIZE`

nordic-krch commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +279,40 @@ config LOG_FRONTEND
+	  In other case, when memory is full, new messages are dropped.
+
+config LOG_FLASH_FILE_SIZE
+	int "Log file size as multiplication of flash page size."
```
is it possible that file system is build on medium without pages (ram?) or backend relies on file system and flash api?

nordic-krch commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +279,40 @@ config LOG_FRONTEND
+	  Max log file size (in bytes) is a multiplication of
+	  LOG_FLASH_FILE_SIZE and flash page size.
+
+config LOG_FLASH_FILENAME
```
maybe `FILE_PREFIX`?

MSyc-NS replied:
Renamed to `LOG_BACKEND_FS_FILE_PREFIX`.

nordic-krch commented on subsys/logging/log_backend_flash.c:
```diff
@@ -0,0 +1,222 @@
+	}
+	rc = fs_write(f, data, length);
+	if (rc == LFS_ERR_NOSPC) {
+#if defined(CONFIG_OVERWRITE_OLD_LOG_FILES)
```
use `if (IS_ENABLED(CONFIG_OVERWRITE_OLD_LOG_FILES))` instead. This way ifdefed content is compiled. You can then remove ifdef from `delete_oldest_logs` declaration. Compiler will get rid of things which are determined to be unused during compilation.

nordic-krch commented on subsys/logging/log_backend_flash.c:
```diff
@@ -0,0 +1,222 @@
+
+	rc = fs_mount(mp);
+	if (rc < 0) {
+		return;
```
you probably need some kind of state flag which will record failed initialization and skip any attempt to write logs.

MSyc-NS replied:
Introduced `backend_state` flag indicating current backend state.

nordic-krch commented on subsys/logging/log_backend_flash.c:
```diff
@@ -0,0 +1,222 @@
+	}
+	(void)fs_closedir(&dir);
+
+#if defined(CONFIG_OVERWRITE_OLD_LOG_FILES)
```
ditto

nordic-krch commented on subsys/logging/log_backend_flash.c:
```diff
@@ -0,0 +1,222 @@
+	fs_statvfs(lfs_storage_mnt.mnt_point, &stat);
+
+	if (stat.f_bfree <= CONFIG_LOG_FLASH_FILE_SIZE) {
+#if defined(CONFIG_OVERWRITE_OLD_LOG_FILES)
```
ditto

nordic-krch commented on subsys/logging/log_backend_flash.c:
```diff
@@ -0,0 +1,222 @@
+		return length;
+#endif
+	}
+	if (rc < 0) {
```
else if

nordic-krch commented on subsys/logging/log_backend_flash.c:
```diff
@@ -0,0 +1,222 @@
+	if (rc < 0) {
+		return 0;
+	}
+	fs_sync(f);
```
usually new line is added after closing bracket.

nordic-krch commented on subsys/logging/log_backend_flash.c:
```diff
@@ -0,0 +1,222 @@
+		return 0;
+	}
+	fs_sync(f);
+	return length;
```
usually new line is added before return statement.

nordic-krch reviewed (commented):
few comments, Have you consider adding some kind of test to verify that?

nvlsianpu commented on subsys/logging/log_backend_flash.c:
```diff
@@ -0,0 +1,222 @@
+	struct fs_mount_t *mp = &lfs_storage_mnt;
+	int file_num = 0;
+
+	rc = fs_mount(mp);
```
I think FS mounting should be optional - application might mount FS somewhere else and somwhen else as well.

nvlsianpu replied:
I think we should have in zephyr post kernel initialization step where FS can be mounted automatically- this is different task for sure.
@nordic-krch what do you think about my comments?

MSyc-NS replied:
@nvlsianpu If FS mounting should be optional, how backend could get know about the one mounted by application? Should it be provided by any special API?

nvlsianpu replied:
Backend should use only proper file/directory name provide via Kconfig (as is here).

nvlsianpu replied:
What with this?

MSyc-NS replied:
Mount point is now provided via Kconfig.

nvlsianpu commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +279,40 @@ config LOG_FRONTEND
 
+config LOG_BACKEND_FLASH
+	bool "Enable FLASH backend"
+	depends on FLASH
```
```suggestion
	depends on FILE_SYSTEM
```

This patch shouldn't depends on FS implementation but on VFS API avielable.

nvlsianpu commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,240 @@
+static enum backend_fs_state backend_state = BACKEND_FS_NOT_INITIALIZED;
+
+static int allocate_new_file(struct fs_file_t *file);
+static int delete_oldest_logs(void);
```
```suggestion
#if CONFIG_LOG_BACKEND_FS_OVERWRITE
static int delete_oldest_logs(void);
#endif
```

nvlsianpu reviewed (changes requested):
I think there is need to limit how much memory can occupy logger. It can be done by limit on spare memory or limit on numbers of logger files.
Otherwise it will always have a tendency to take up all available memory, which might conflict with other FS clients,

Commit:
susbys: logging: Added logging to flash backend

New backend is based on littleFS. After init, there is created new file
with continuous numeration. When max size of file is reached, system
creates another one. File size is limited by Kconfig option.
There is possibility to overwrite old files or drop new ones.
FS backend logging to file if the FS location is available.
Otherwise log messages are drooped.

User can also change the name of log files.

Signed-off-by: Mateusz Syc <Mateusz.Syc@nordicsemi.no>
Signed-off-by: Andrzej Puzdrowski <Andrzej.Puzdrowski@nordicsemi.no>

MSyc-NS commented:
> I think there is need to limit how much memory can occupy logger. It can be done by limit on spare memory or limit on numbers of logger files.
> Otherwise it will always have a tendency to take up all available memory, which might conflict with other FS clients,

Number of files is now limited by `CONFIG_LOG_BACKEND_FS_FILES_LIMIT`. Connecting with `CONFIG_LOG_BACKEND_FS_FILE_SIZE` gives user control over space dedicated to log files.

nvlsianpu reviewed (commented):
Implementation looks fine now, just one nip.

nvlsianpu reviewed (commented):
Can test of file switching, file overwriting, file size limit,  file numbers limit be added?

de-nordic commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,254 @@
+
+	sprintf(fname, "%s/%s%04d", CONFIG_LOG_BACKEND_FS_MOUNT_POINT,
+		CONFIG_LOG_BACKEND_FS_FILE_PREFIX, file_num);
+	fs_statvfs(CONFIG_LOG_BACKEND_FS_MOUNT_POINT, &stat);
```
What if `fs_statvfs` fails to fill in `stat`, the loop will be processing incorrect data.

de-nordic commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,254 @@
+		CONFIG_LOG_BACKEND_FS_FILE_PREFIX, file_num);
+	fs_statvfs(CONFIG_LOG_BACKEND_FS_MOUNT_POINT, &stat);
+
+	while ((file_ctr >= CONFIG_LOG_BACKEND_FS_FILES_LIMIT) ||
```
Since you want to call `fs_statvfs` at least once, then you may probably  do the `do { ... } while (...) loop instead.

MSyc-NS replied:
Yes, I want to call `fs_statvfs` at least once, but not want to `delete_oldest_logs()` if there is no need. In your proposition to avoid this, I should insert additional `if` inside `do...while` loop.

de-nordic commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,254 @@
+		}
+	}
+
+	rc = fs_open(file, fname, FS_O_CREATE);
```
You are missing `FS_O_WRITE` flag here, if you are opening a file for write. The fact that you have been able to write the file without the flag comes from collateral effect of `CREATE` flag within LittleFS and is not warranted on other operating systems.

de-nordic commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,254 @@
+
+			sprintf(del_file_path, "%s/%s",
+				CONFIG_LOG_BACKEND_FS_MOUNT_POINT, ent.name);
+			rc = fs_unlink(del_file_path);
+			if (rc < 0) {
+				break;
+			}
+			file_ctr--;
+			break;
```
`fs_util`  may return `-ENOENT` for example, which means that the file has already been removed (somehow), is it an error then?

I think this code should be moved out of `while` loop. You are processing one file only so the loop should be dedicated to identifying it and the code that operates on it should be outside.

MSyc-NS replied:
I considered this case and I think dealing with `-ENOENT` error should be the similar to properly deleted file. 

As you proposed I moved `fs_unlink` part of code outside loop.

de-nordic reviewed (changes requested):
Some comments requiring discussion.

MSyc-NS commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,254 @@
+		fs_close(file);
+		char *name = strstr(fname, CONFIG_LOG_BACKEND_FS_FILE_PREFIX);
+
+		file_num = atoi(name + LOG_PREFIX_LEN) + 1;
```
I wanted to avoid the situation, when user or another activity artificially creates new log file in the runtime. Basic incrementation cannot catch this case; a new file will be always one after last.

de-nordic replied:
ok, reasonable argument.

MSyc-NS commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,254 @@
+			fs_statvfs(CONFIG_LOG_BACKEND_FS_MOUNT_POINT, &stat);
+			if (rc < 0) {
+				return rc;
+			}
```
You are right. I insterted additional checks of `fs_statvfs` and  error handling. Endless loop should not occur.

de-nordic commented:
I think the entire feature should be renamed "File System backed log" because it operates on file systems not directly on flash.

de-nordic commented on tests/subsys/logging/log_backend_fs/src/log_fs_test.c:
```diff
@@ -0,0 +1,126 @@
+		struct fs_dirent ent = { 0 };
+
+		rc = fs_readdir(&dir, &ent);
+		if (rc < 0) {
+			break;
+		}
+		if (ent.name[0] == 0) {
+			break;
+		}
```
You can bundle both expressions in one if.

de-nordic commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,273 @@
+			struct fs_dirent ent = { 0 };
+
+			rc = fs_readdir(&dir, &ent);
+			if (rc < 0) {
+				break;
+			}
+			if (ent.name[0] == 0) {
+				break;
+			}
```
You can make it one `if` with `||`.

nordic-krch commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,276 @@
+
+enum backend_fs_state {
+	BACKEND_FS_OK = 0,
+	BACKEND_FS_CORUPTED,
```
CORRUPTED

nordic-krch commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,252 @@
+#include "log_backend_std.h"
+#include <assert.h>
+#include <fs/fs.h>
+#include <fs/littlefs.h>
```
why do you want to ifdef header? That prevents use of `if(IS_ENABLED(CONFIG_LOG_BACKEND_FS_AUTOSTART)`

nvlsianpu replied:
^^@nordic-krch please clarify your thought.

nvlsianpu replied:
clarified off-channel, @nordic-krch has right

nordic-krch commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,276 @@
+	int rc;
+#if CONFIG_LOG_BACKEND_FS_AUTOSTART
+	/* If autostart, mount file system automatically. */
+	struct fs_mount_t *mp = &lfs_storage_mnt;
```
ideally you could use:
```
if(IS_ENABLED(CONFIG_LOG_BACKEND_FS_AUTOSTART) {
...
}
```
and avoid ifdefs here and in struct declaration.

MSyc-NS replied:
Is it preferred to use `if(IS_ENABLED(...))` than `#if ...`? In the second option `static struct fs_mount lfs_storage_mnt` is not compiled (less size).

nordic-krch replied:
in both cases it won't be placed in rom/ram. Advantage of `if (IS_ENABLED()` is that things get compiled. If variable is static and not used in the file it will be removed. After preprocessing compiler gets:
```
static int foo;

if (0) {
  foo = ...
}

if (0) {
  if (foo) ...
}
```
compiler figures out that foo is never used and removes it.

nordic-krch commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,276 @@
+
+static int allocate_new_file(struct fs_file_t *file);
+
+#if CONFIG_LOG_BACKEND_FS_OVERWRITE
```
please remove it, otherwise compilation will fail when overwriting is disabled.

nordic-krch commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,276 @@
+enum backend_fs_state {
+	BACKEND_FS_OK = 0,
+	BACKEND_FS_CORUPTED,
+	BACKEND_FS_NOT_INITIALIZED
```
i would have it as 0 so that `backend_state` can land in zero initialize section.

nordic-krch commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,276 @@
+
+static int allocate_new_file(struct fs_file_t *file)
+{
+	printf("Allocate new file\n");
```
remove this

nordic-krch commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,276 @@
+
+static int write_log_to_file(uint8_t *data, size_t length, void *ctx)
+{
+	if (backend_state == BACKEND_FS_NOT_INITIALIZED) {
```
is this state even possible? At the end of `log_backend_fs_init()` `allocate_new_file` is called.

MSyc-NS replied:
I missed that user should not only enable backend, but invoke `api->init()` too. Therefore I put initialization in `write_log_to_file()`. Will fix it.

nordic-krch commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,276 @@
+		int size = fs_tell(f);
+
+		if (size < 0) {
+			backend_state = BACKEND_FS_CORUPTED;
```
what about adding here (and in other premature terminations) `goto on_error;` 
and add label at the end of the function:
```
on_error:
    backend_state = BACKEND_FS_CORUPTED;
    return length;
```

nashif commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,276 @@
+
+static int allocate_new_file(struct fs_file_t *file)
+{
+	printf("Allocate new file\n");
```
debug messages?

nvlsianpu replied:
removed.

nordic-krch commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,276 @@
+		}
+	}
+	if (del_file_path[0] != '\0') {
+		rc = fs_unlink(del_file_path);
```
could it be moved right after `sprintf(del_file_path,..`?

MSyc-NS replied:
Earlier it was inside `while` loop. I replaced it as @de-nordic said in [this](https://github.com/zephyrproject-rtos/zephyr/pull/27486#discussion_r477238209) comment:

> I think this code should be moved out of while loop. You are processing one file only so the loop should be dedicated to identifying it and the code that operates on it should be outside.

nordic-krch reviewed (commented):
few comments

nordic-krch commented on subsys/logging/CMakeLists.txt:
```diff
@@ -14,6 +14,15 @@ if(NOT CONFIG_LOG_MINIMAL)
+  )
+
+  zephyr_include_directories(
+    ${ZEPHYR_BASE}/../modules/fs/littlefs
```
is it needed here. doesn't it come with littlefs being enabled?

MSyc-NS replied:
There is an error without above code inserted and `CONFIG_FILE_SYSTEM_LITTLEFS=y` set: `lfs.h No such file or directory`.

nordic-krch replied:
still, if there is something wrong it should not be solved here.
There is a `CMakeLists.txt` in `modules/fs/littlefs`. Maybe it is something wrong there?
```
target_include_directories(LITTLEFS INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
```
@pabigot as you were the one adding it, do you have any ideas?

pabigot replied:
The file systems are packaged into libraries that provide all the flags and options required to make use of them.

Following the shell subsystem I think the correct fix is to add something like this:

  target_link_libraries_ifdef(CONFIG_FAT_FILESYSTEM_ELM   LOG INTERFACE ELMFAT)
  target_link_libraries_ifdef(CONFIG_FILE_SYSTEM_LITTLEFS LOG INTERFACE LITTLEFS)

but it appears that requires making the log subsystem an interface library like the shell subsystem is.  There may be a better solution; a CMake expert like @tejlmand can provide more informed help.

nordic-krch commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +280,58 @@ config LOG_FRONTEND
+	bool "Enable file system backend start automatically"
+	default y
+	help
+	  Allow backend to start autmatically, initialize it's own
```
autmatically->automatically
shouldn't it mention that littlefs is being is used?

tejlmand commented on subsys/logging/CMakeLists.txt:
```diff
@@ -14,6 +14,15 @@ if(NOT CONFIG_LOG_MINIMAL)
+    log_backend_fs.c
+  )
+
+  zephyr_include_directories(
+    ${ZEPHYR_BASE}/../modules/fs/littlefs
+  )
```
```suggestion
  if(CONFIG_LOG_BACKEND_FS AND CONFIG_FILE_SYSTEM_LITTLEFS)
    zephyr_link_libraries(LITTLEFS)
  endif()
```

The reason for the `CONFIG_FILE_SYSTEM_LITTLEFS` test was because I could only see `LOG_BACKEND` is dependent on `FILE_SYSTEM` and not a specific `FILE_SYSTEM` implementation.
https://github.com/zephyrproject-rtos/zephyr/blob/9f7c0f3c0ef399ca5afae486aca19e694bda6eac/subsys/logging/Kconfig#L285

tejlmand replied:
Please let me know if this doesn't work.

tejlmand replied:
if making `LOG_BACKEND_FS` depends on `FILE_SYSTEM_LITTLEFS`, then this can be simplified further into:

```suggestion
  zephyr_link_libraries_ifdef(CONFIG_LOG_BACKEND_FS LITTLEFS)
```

MSyc-NS replied:
> Please let me know if this doesn't work.

It works fine. I used the last version
`zephyr_link_libraries_ifdef(CONFIG_LOG_BACKEND_FS LITTLEFS)`
as the LittleFS is the only currently tested FS.

tejlmand commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +280,58 @@ config LOG_FRONTEND
 
+config LOG_BACKEND_FS
+	bool "Enable file system backend"
+	depends on FILE_SYSTEM
```
Has this PR been tested with anything else than littlefs ?

tejlmand replied:
should probably be, based on the description of this PR https://github.com/zephyrproject-rtos/zephyr/pull/27486#issue-465992198.
> This PR introduces a new backend based on littleFS.

```suggestion
	depends on FILE_SYSTEM_LITTLEFS
```

MSyc-NS replied:
Replaced.

tejlmand commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +280,58 @@ config LOG_FRONTEND
+	bool "Enable file system backend start automatically"
+	default y
+	help
+	  Allow backend to start automatically. Initialize it's own
+	  LittleFS instance and mount it.
```
If only LittleFS is supported, then that should be made very clear.

MSyc-NS replied:
I replaced `Enable LittleFS backend` as description of `LOG_BACKEND_FS` in Kconfig. Your proposition of `depends on FILE_SYSTEM_LITTLEFS` was also inserted.

Commit:
samples: subsys: mgmt: smp_svr: Possible download log file from flash FS

Added ability to downlaod log file from flash with LittleFS
using mcumgr.

Signed-off-by: Mateusz Syc <Mateusz.Syc@nordicsemi.no>

Commit:
tests/subsys/logging/log_backend_fs: Test of log_backend_fs

Tests verify correctness of the internal backend behaviour.

Signed-off-by: Mateusz Syc <Mateusz.Syc@nordicsemi.no>
Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

pabigot commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,263 @@
+		if (rc < 0) {
+			break;
+		}
+		if (strstr(ent.name, CONFIG_LOG_BACKEND_FS_FILE_PREFIX) !=
```
This seems fragile.  What guarantee is there that the dirent order is oldest first?  This should extract the sequence number from the file name and use the numerically smallest one found.

It should also deal with sequence number wraparound if that's going to be handled.

pabigot commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,263 @@
+	.mnt_point = CONFIG_LOG_BACKEND_FS_MOUNT_POINT,
+};
+
+static char fname[MAX_PATH_LEN];
```
You could save some RAM here by only reserving the amount required for the prefix and the sequence number,

pabigot commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,263 @@
+	}
+
+	sprintf(fname, "%s/%s%04d", CONFIG_LOG_BACKEND_FS_MOUNT_POINT,
+		CONFIG_LOG_BACKEND_FS_FILE_PREFIX, file_num);
```
When `file_num` exceeds 9999 it won't be a 4-digit sequence anymore.  I guess that'll work as long as you use the full `MAX_PATH_LEN` (probably 256) as the storage buffer, though that's overkill.

This should also use snprintf().

pabigot commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +280,58 @@ config LOG_FRONTEND
+
+config LOG_BACKEND_FS_AUTOSTART
+	bool "Enable file system backend start automatically"
+	default y
```
IIRC there was some other backend (net?) that recently had to be made not auto-start.  Is this safe when the file system is on an external flash device that isn't available until late `POST_KERNEL`?

pabigot reviewed (changes requested):
I need this capability, so I'd like to see it go in.  OTOH there are some things that need to be fixed: specifically the `sprintf` and the assumption that directory order is oldest first.  Those could be done as bug fixes after the merge if there's a risk this won't make it in.

pabigot commented:
To clarify: I will remove my NAK if that allows this to get in to 2.4, but somebody will need to fix the issues I've identified before release.  As I will be using this I can say that fixes will be made available.

de-nordic commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,263 @@
+		/* If autostart, mount file system automatically. */
+		struct fs_mount_t *mp = &lfs_storage_mnt;
+
+		rc = fs_mount(mp);
```
@nvlsianpu @pabigot Question regarding this mount, the LittleFS module gets initialized with:
`SYS_INIT(littlefs_init, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);`
where default for `CONFIG_KERNEL_INIT_PRIORITY_DEFAULT` is 40; the  logger_core  is initialized with:
`SYS_INIT(enable_logger, POST_KERNEL, 0);`
doesn't it mean that littlefs gets initialized **after** this module?

pabigot replied:
Thanks for confirming; yes, that seems to be the case, which means it might have problems with autostart.

aunsbjerg commented:
Did some digging on the issues I've have with UDP + FS for the smp_srv sample. When both overlays are added, execution stops after `*** Booting Zephyr ***` is printed but before anything else is printed. Setting `CONFIG_NET_CONFIG_SETTINGS=n` and running `net_config_init_app()` _after_ mounting the FS let's the code run... once. Also, `CONFIG_MPU_ALLOW_FLASH_WRITE=y` is necessary for `frdm_k64f`. 

I don't have time to dig further and I'm not even sure if this is an issue with the network stack, fs logging backend or a combination.

pabigot commented on subsys/logging/log_backend_fs.c:
```diff
@@ -0,0 +1,263 @@
+
+const struct log_backend_api log_backend_fs_api = {
+	.put = put,
+	.put_sync_string = NULL,
```
Should there be a build-assert here if `!IS_ENABLED(CONFIG_LOG_IMMEDIATE)`?  The configuration here seems to require that selection.

pabigot reviewed (changes requested):
Some functional and style comments.  Based on a test using #28086 I can confirm that autostarting this is not going to work with a filesystem on external flash.

Also the logging infrastructure seems unprepared for a backend that cannot be initialized at the same time as every other backend during startup.  `log_backend_activate()` doesn't provide a way to tell the backend that it's being activated, which is where the fs backend could locate and mount the file system.  A solution to that needs to be identified.

nvlsianpu commented on subsys/logging/Kconfig:
```diff
@@ -279,6 +280,58 @@ config LOG_FRONTEND
+config LOG_BACKEND_FS_FILE_SIZE
+	int "User defined log file size"
+	default 4096
+	range 128 16384
```
@pabigot What about range removal?

pabigot replied:
That should be fine.

nvlsianpu commented:
In my opinion backend autostart should based on outstanding feature of:
https://github.com/zephyrproject-rtos/zephyr/pull/28320

pabigot commented:
> In my opinion backend autostart should based on outstanding feature of: #28320

I'm not entirely convinced that the fs backend should be automatically started, but rather should allow retroactive enable like network logging does.  Logging is normally enabled first thing post-kernel.  File systems normally start after that.  If autostart is going to require delaying logging until after the file system is up and running, isn't there a chance of losing information that could have been conveyed through other log interfaces?

I'm not going to insist on this, but I do think revising the log system API to allow enabling this backend later would be more flexible than relying on a filesystem automount.

optdcw commented:
Hello, 
I think we should consider extraction of the feature for rotating file to separate module. I think that rotating file is feature which could be helpful in other cases.

pabigot commented:
> I think we should consider extraction of the feature for rotating file to separate module. I think that rotating file is feature which could be helpful in other cases.

Agreed; I have in mind a service that different components can register with, that can be invoked when space drops below a threshold and allows the components to decide which of their files can be removed.

But even when that's available this needs to work when global auto-rotation isn't enabled, and having a set of N files each of which is capped at M bytes is a good solution as long as the names are maintained so the logs aren't overwritten every time the system reboots.

nvlsianpu commented:
> I think we should consider extraction of the feature for rotating file to separate module. I think that rotating file is feature which could be helpful in other cases.

I have no objection, just prefer iterative approach, so can be done in subsequent patch.

koffes commented:
@MSyc-NS Thanks for the PR. Would it be much work to make this file system agnostic? I can imagine logs being written to flash or an SD-card and someone wants to mount this on a PC. In which case FAT has more support. This can always be added later.

nvlsianpu commented:
> Would it be much work to make this file system agnostic? 

@koffes

This is the plan. I awaiting for #28320 which will provide auto-FS mount. Also as far I know ther is even solution for mounting LittleFS volume under linux.

pabigot commented:
See also #28892.  We should only have one implementation, we should look at both to compare features.

nvlsianpu commented:
@pabigot I see place for both of them. I have been receiving interest on either FS backend and simpler storage backend.

nvlsianpu committed:
subsys/loggin/log_backend_fs: added recovery after file lost

Added recovery for case when log file was lost somehow.
Back-end will try to create new file if possible.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
logging/log_backend_fs: support any directory

Added procedure which is able to create nested log director.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
superseded by #32973
